pkg/util/slice_operations.go

```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package util
+
+import (
+	"sync"
+)
+
+type idxRange struct {
+	from int
+	to   int
+}
+
+var idxRangesPool = sync.Pool{
+	New: func() interface{} {
+		return &[]idxRange{}
+	},
+}
+
+// RemoveSliceIndexes takes a slice of elements of any type and a sorted slice of indexes of elements in the former slice.
+// It removes the elements at the given indexes from the given slice, while minimizing the number of copy operations
+// that are required to do so by grouping consecutive indexes into ranges and removing the ranges in single operations.
+// The given number of indexes must be sorted in ascending order.
+// Indexes which are duplicate or out of the range of the given slice are ignored.
+// The returned values are:
+// - The resulting slice of elements after the elements at the given indexes have been removed.
+// - The number of elements that have been removed.
+// - The number of ranges of consecutive elements that have been removed.
+func RemoveSliceIndexes[T any](data []T, indexes []int) ([]T, int, int) {
+	ranges := idxRangesPool.Get().(*[]idxRange)
+	none := idxRange{-1, -1}
+	currentRange := none
+
+	// Keep track of the last index to filter duplicates.
+	lastIndex := -1
+	for _, index := range indexes {
+		if index < 0 || index >= len(data) {
+			// Invalid indexes given, ignore them.
+			continue
+		}
+
+		// Skip duplicates.
+		if index == lastIndex {
+			continue
+		}
+		lastIndex = index
+
+		// Initialize the current range variable.
+		if currentRange == none {
+			currentRange.from = index
+			currentRange.to = index
+			continue
+		}
+
+		// Extend the current range by one, because this index connected to it.
+		if currentRange.to+1 == index {
+			currentRange.to = index
+			continue
+		}
+
+		// The current range is finished, add it to the ranges list.
+		*ranges = append(*ranges, currentRange)
+
```

Comment:
Why do we build list of ranges first, and don't just do the removal at this point?

Reply:
Implemented in https://github.com/grafana/mimir/pull/2703/commits/12a98bf242b0da30acd287a17cc0fbc2a43b5f59.

Reply:
you're right, building the list of ranges makes no sense 🤦‍♂️